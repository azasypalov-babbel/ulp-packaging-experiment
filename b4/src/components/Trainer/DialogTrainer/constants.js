export const TIMEOUT_AFTER_CONTEXTUAL_INFO = 2000;
