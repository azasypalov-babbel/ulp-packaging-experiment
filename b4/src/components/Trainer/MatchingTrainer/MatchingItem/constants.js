import { rem } from 'polished';

export const APPEARANCE_TYPE = {
  DEFAULT: 'DEFAULT',
  ACTIVE: 'ACTIVE',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR'
};

export const itemPairMargin = rem('16px');
export const horizontalTravel = '6rem';
