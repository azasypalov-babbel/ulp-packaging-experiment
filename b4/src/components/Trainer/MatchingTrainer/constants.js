export const MATCHING_TRAINER_STATE = {
  READY: 'READY',
  PLAYING_AUDIO: 'PLAYING_AUDIO',
  AWAITING_CONTINUE: 'AWAITING_CONTINUE',
  DONE: 'DONE',
  FEEDBACK: 'FEEDBACK'
};

export const ITEM_FRAGMENT_TYPE = {
  BASE: 'BASE',
  OPTION: 'OPTION'
};

export const trainerMaxWidth = '56rem';
