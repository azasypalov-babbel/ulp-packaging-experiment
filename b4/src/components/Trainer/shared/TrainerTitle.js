import React from 'react';
import PropTypes from 'prop-types';

import { rem } from 'polished';
import styled from 'styled-components';
import { markupStringToHTML } from '@lessonnine/babbel-markup-helper.js';

import Title from '../../shared/Title';

const StyledTitle = styled(Title)`
  margin-left: 2.25rem;
  margin-right: 2.25rem;
  margin-bottom: ${rem(24)};

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xxsmall}) {
    margin-bottom: ${rem(32)};
  }

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    margin-bottom: ${rem(48)};
  }
`;

const TrainerTitle = ({ text, ...props }) => (
  <StyledTitle
    data-selector="title"
    textAlign="center"
    fontFamily="fontMilliardSemi"
    color="spaceGray"
    fontSize="big"
    as="h1"
    dangerouslySetInnerHTML={{ __html: markupStringToHTML(text) }}
    {...props}
  />
);

TrainerTitle.propTypes = {
  text: PropTypes.string.isRequired
};

export default TrainerTitle;
