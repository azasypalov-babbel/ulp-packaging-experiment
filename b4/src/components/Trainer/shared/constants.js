export const PLAYBACK_RATE_NORMAL = 1;
export const PLAYBACK_RATE_SLOW = 0.7;
