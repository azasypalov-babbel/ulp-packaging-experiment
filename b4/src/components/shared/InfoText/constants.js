export const TIMEOUT_AFTER_INFOTEXT_DISPLAY = 1000;
