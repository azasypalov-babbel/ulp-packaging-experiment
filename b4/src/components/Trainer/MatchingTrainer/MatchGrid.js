import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import MatchingItem from './MatchingItem/MatchingItem';
import { PoseGroup } from 'react-pose';
import { scrollIntoSafeArea } from '../../shared/scrollIntoSafeArea';
import { ITEM_FRAGMENT_TYPE, MATCHING_TRAINER_STATE } from './constants';
import { APPEARANCE_TYPE, itemPairMargin } from './MatchingItem/constants';
import useBreakpoint from '../../shared/hooks/useBreakpoint';
import { matchGridPropType } from './useMatchingGridState';

const StyledPairListWrapper = styled.div`
  display: flex;
  flex-direction: column;

  flex: 1 1 100%;

  scroll-padding: ${itemPairMargin};

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    display: grid;
    grid-template-columns: auto auto;
    grid-auto-rows: 1fr;

    justify-content: center;

    flex: initial;

    transition: margin 0.25s;
  }
`;

const useOnItemChange = ({ onChange, grid }) => {
  const activeItem = useRef(null);
  useEffect(() => {
    const active = grid.find(({ isActive }) => isActive);
    if (active?.id && active.id !== activeItem.current) {
      activeItem.current = active.id;
      onChange();
    }
  }, [grid, onChange]);
};

const Spacer = styled.div`
  display: block;
  order: 0;
  flex: 1 0 4rem;

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    display: none;
  }
`;

// eslint-disable-next-line no-unused-vars
const getAppearance = (isAllMatched, isGivingFeedback, { isActive, isSuccess, isError }) => {
  if (isAllMatched) return APPEARANCE_TYPE.DEFAULT;

  if (isGivingFeedback) {
    if (isSuccess) return APPEARANCE_TYPE.SUCCESS;
    if (isError) return APPEARANCE_TYPE.ERROR;
  }
  if (isActive) return APPEARANCE_TYPE.ACTIVE;
  return APPEARANCE_TYPE.DEFAULT;
};

const breakpoints = ['xxxsmall', 'xsmall'];

const shouldHideOnMobileLayout = ({ isSuccess, isError, isHidden, isMatched, isActive }) => {
  if (isHidden) return true;

  if (isSuccess || isError) return false;
  if (isActive) return false;

  // Items become matched as soon as they're correct
  // We want to still see the items while feedback is being given (isSuccess, isError, isActive)
  // That is why this is checked last
  if (isMatched) return true;

  return false;
};

const MatchGrid = ({ grid, onOptionClick, trainerState, showTranslation }) => {
  const matchGrid = useRef(null);

  useOnItemChange({
    onChange: () => { scrollIntoSafeArea(matchGrid.current); },
    grid
  });

  const activeBreakpoint = useBreakpoint(breakpoints);

  const isAllMatched = trainerState === MATCHING_TRAINER_STATE.AWAITING_CONTINUE;
  const isAwaitingSoundPlay = trainerState === MATCHING_TRAINER_STATE.PLAYING_AUDIO;
  const isDone = trainerState === MATCHING_TRAINER_STATE.DONE;
  const isGivingFeedback = trainerState === MATCHING_TRAINER_STATE.FEEDBACK;
  const isMobileLayout = activeBreakpoint === breakpoints[0];

  return (
    <StyledPairListWrapper
      ref={matchGrid}
    >
      <PoseGroup flipMove={!isAllMatched}>
        {grid
          // Filter out hidden items on mobile breakpoint
          .filter((item) => {
            if (!isMobileLayout) return true;
            if (isAllMatched) return true;
            if (isDone) return false;
            return !shouldHideOnMobileLayout(item);
          })
          .map((itemFragment, index, list) => {
            const shouldInvertTooltipPosition = index === list.length - 1;
            const { correspondingBaseId, id, isActive, isHidden, isMatched, soundId, text, translation, type} = itemFragment;
            const isOption = type === ITEM_FRAGMENT_TYPE.OPTION;
            const appearance = getAppearance(
              isAllMatched,
              isGivingFeedback,
              itemFragment
            );
            const order = isMatched || isActive ? -1 : 1;

            return (
              <MatchingItem
                data-solution-id={isOption ? correspondingBaseId : id}
                key={id + type}
                appearance={appearance}
                isClickable={!isGivingFeedback && !isAwaitingSoundPlay}
                id={id}
                isHidden={isHidden}
                isMatched={isMatched}
                isOption={isOption}
                itemText={text.toHTML()}
                onOptionClick={onOptionClick}
                order={order}
                position={type}
                tooltipPosition={shouldInvertTooltipPosition ? "TOP" : "BOTTOM"}
                showTranslation={showTranslation}
                soundId={soundId}
                trainerActive={!isAllMatched}
                translation={translation?.toHTML()}
              />
            );
          })}
      </PoseGroup>
      { !isAllMatched && <Spacer /> }
    </StyledPairListWrapper>
  );
};

MatchGrid.propTypes = {
  onOptionClick: PropTypes.func.isRequired,
  trainerState: PropTypes.oneOf(Object.values(MATCHING_TRAINER_STATE)).isRequired,
  grid: matchGridPropType.isRequired,
  showTranslation: PropTypes.bool.isRequired
};

export default MatchGrid;
