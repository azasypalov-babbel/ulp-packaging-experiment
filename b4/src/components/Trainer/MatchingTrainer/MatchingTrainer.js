import React from 'react';
import StyledTrainerLayout from '../../TrainerLayout';
import PropTypes from 'prop-types';
import MatchGrid from './MatchGrid';
import styled from 'styled-components';
import withTranslations from '../../shared/withTranslations';
import TrainerTitle from '../shared/TrainerTitle';
import { rem } from 'polished';
import { MATCHING_TRAINER_STATE } from './constants';
import { matchGridPropType } from './useMatchingGridState';

const StyledMatchingTrainer = styled(StyledTrainerLayout)`
  display: flex;
  flex: 1 1 100%;
  flex-direction: column;

  text-align: center;
  margin-left: ${rem(24)};
  margin-right: ${rem(24)};
  margin-bottom: calc(1rem + env(safe-area-inset-bottom));

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    margin-left: ${rem(64)};
    margin-right: ${rem(64)};
  }
`;

export const MatchingTrainer = ({
  trainerState,
  grid,
  title,
  translations,
  onOptionClick,
  showTranslation
}) => {
  return (
    <StyledMatchingTrainer>
      <TrainerTitle text={title || translations.trainerTitle} />
      <MatchGrid
        trainerState={trainerState}
        grid={grid}
        onOptionClick={onOptionClick}
        showTranslation={showTranslation}
      />
    </StyledMatchingTrainer>
  );
};

const getTranslations = (translate) => {
  return {
    trainerTitle: translate('b3.page_matching.match_up_the_correct_items')
  };
};

MatchingTrainer.propTypes = {
  title: PropTypes.string,
  trainerState: PropTypes.oneOf(Object.values(MATCHING_TRAINER_STATE)).isRequired,
  grid: matchGridPropType.isRequired,
  translations: PropTypes.shape({
    trainerTitle: PropTypes.string
  }),
  onOptionClick: PropTypes.func.isRequired,
  showTranslation: PropTypes.bool.isRequired
};

export default withTranslations(getTranslations)(MatchingTrainer);
