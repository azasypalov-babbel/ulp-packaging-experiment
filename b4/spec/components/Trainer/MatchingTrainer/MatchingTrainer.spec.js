import React from 'react';
import { shallow } from 'enzyme';
import { MatchingTrainer } from '../../../../src/components/Trainer/MatchingTrainer/MatchingTrainer';
import TrainerTitle from '../../../../src/components/Trainer/shared/TrainerTitle';
import MatchGrid from '../../../../src/components/Trainer/MatchingTrainer/MatchGrid';
import { ITEM_FRAGMENT_TYPE } from '../../../../src/components/Trainer/MatchingTrainer/constants';
import { MATCHING_TRAINER_STATE } from '../../../../src/components/Trainer/MatchingTrainer/constants';

jest.mock('../../../../src/components/shared/hooks/useBreakpoint');

const defaultProps = {
  title: 'Match it',
  trainerState: MATCHING_TRAINER_STATE.READY,
  grid: [
    {
      id: '162d2e047d4a2badc81e00aa574b15e0',
      text: { children: [], isLeaf: false, toHTML: jest.fn(() => 'the fruit') },
      isActive: true,
      isHidden: false,
      isMatched: false,
      type: ITEM_FRAGMENT_TYPE.BASE
    },
    {
      id: 0,
      correspondingBaseId: '162d2e047d4a2badc81e00aa574b15e0',
      text: { children: [], isLeaf: false, toHTML: jest.fn(() => 'das Obst') },
      translation: { children: [], isLeaf: false, toHTML: jest.fn(() => 'translation') },
      isSuccess: false,
      isMatched: false,
      isError: false,
      type: ITEM_FRAGMENT_TYPE.OPTION
    },
    {
      id: '33df6f3aa74afbe4d6f541825f6e758d',
      text: { children: [], isLeaf: false, toHTML: jest.fn(() => 'the cheese') },
      isActive: false,
      isHidden: false,
      isMatched: false,
      type: ITEM_FRAGMENT_TYPE.BASE
    },
    {
      id: 1,
      correspondingBaseId: '33df6f3aa74afbe4d6f541825f6e758d',
      text: { children: [], isLeaf: false, toHTML: jest.fn(() => 'der KÃ¤se') },
      translation: { children: [], isLeaf: false, toHTML: jest.fn(() => 'translation cheese') },
      isSuccess: false,
      isMatched: false,
      isError: false,
      type: ITEM_FRAGMENT_TYPE.OPTION
    },
    {
      id: 'b9cf73d37e1dd788faa5bf3cfaf3fd44',
      text: { children: [], isLeaf: false, toHTML: jest.fn(() => 'the rhino') },
      isActive: false,
      isHidden: false,
      isMatched: false,
      type: ITEM_FRAGMENT_TYPE.BASE
    },
    {
      id: 2,
      correspondingBaseId: 'b9cf73d37e1dd788faa5bf3cfaf3fd44',
      text: { children: [], isLeaf: false, toHTML: jest.fn(() => 'das Rhinozeros') },
      translation: { children: [], isLeaf: false, toHTML: jest.fn(() => 'translation rhino') },
      isSuccess: false,
      isMatched: false,
      isError: false,
      type: ITEM_FRAGMENT_TYPE.OPTION
    }
  ],
  onOptionClick: jest.fn(),
  showTranslation: true
};

describe('<MatchingTrainer />', () => {
  it('renders a title', ()=> {
    const wrapper = shallow(<MatchingTrainer {...defaultProps}  />);
    const title = wrapper.find(TrainerTitle);
    expect(title.exists()).toBe(true);
    expect(title.prop('text')).toEqual('Match it');
  });

  it('renders a matching grid', () => {
    const wrapper = shallow(<MatchingTrainer {...defaultProps}  />);
    const grid = wrapper.find(MatchGrid);
    expect(grid.exists()).toBe(true);
    expect(grid.prop('grid')).toEqual(defaultProps.grid);
    expect(grid.prop('trainerState')).toEqual(defaultProps.trainerState);
    expect(grid.prop('onOptionClick')).toEqual(defaultProps.onOptionClick);
  });
});


describe('<MatchGrid />', () => {
  let wrapper;

  describe('while not giving feedback', () => {
    beforeEach(() => {
      wrapper = shallow(<MatchGrid
        grid={defaultProps.grid}
        trainerState={defaultProps.trainerState}
        onOptionClick={defaultProps.onOptionClick}
        showTranslation={defaultProps.showTranslation}
      />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have clickable options while not giving feedback', () => {
      wrapper.find('WithAnimations(MatchingItem)').forEach((item) => {
        expect(item.prop('isClickable')).toBeTruthy();
      });
    });
  });

  describe('while giving feedback', () => {
    beforeEach(() => {
      wrapper = shallow(<MatchGrid
        grid={defaultProps.grid}
        trainerState={MATCHING_TRAINER_STATE.FEEDBACK}
        onOptionClick={defaultProps.onOptionClick}
        showTranslation={defaultProps.showTranslation}
      />);
    });

    it('should not have a clickable option while giving feedback', () => {
      wrapper.find('WithAnimations(MatchingItem)').forEach((item) => {
        expect(item.prop('isClickable')).toBeFalsy();
      });
    });
  });
});
