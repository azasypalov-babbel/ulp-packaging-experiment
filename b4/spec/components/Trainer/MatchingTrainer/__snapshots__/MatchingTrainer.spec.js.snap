// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MatchGrid /> while not giving feedback should match snapshot 1`] = `
<styled.div>
  <PoseGroup
    flipMove={true}
  >
    <WithAnimations(MatchingItem)
      appearance="ACTIVE"
      data-solution-id="162d2e047d4a2badc81e00aa574b15e0"
      id="162d2e047d4a2badc81e00aa574b15e0"
      isClickable={true}
      isHidden={false}
      isMatched={false}
      isOption={false}
      itemText="the fruit"
      key="162d2e047d4a2badc81e00aa574b15e0BASE"
      onOptionClick={[MockFunction]}
      order={-1}
      position="BASE"
      showTranslation={true}
      tooltipPosition="BOTTOM"
      trainerActive={true}
    />
    <WithAnimations(MatchingItem)
      appearance="DEFAULT"
      data-solution-id="162d2e047d4a2badc81e00aa574b15e0"
      id={0}
      isClickable={true}
      isMatched={false}
      isOption={true}
      itemText="das Obst"
      key="0OPTION"
      onOptionClick={[MockFunction]}
      order={1}
      position="OPTION"
      showTranslation={true}
      tooltipPosition="BOTTOM"
      trainerActive={true}
      translation="translation"
    />
    <WithAnimations(MatchingItem)
      appearance="DEFAULT"
      data-solution-id="33df6f3aa74afbe4d6f541825f6e758d"
      id="33df6f3aa74afbe4d6f541825f6e758d"
      isClickable={true}
      isHidden={false}
      isMatched={false}
      isOption={false}
      itemText="the cheese"
      key="33df6f3aa74afbe4d6f541825f6e758dBASE"
      onOptionClick={[MockFunction]}
      order={1}
      position="BASE"
      showTranslation={true}
      tooltipPosition="BOTTOM"
      trainerActive={true}
    />
    <WithAnimations(MatchingItem)
      appearance="DEFAULT"
      data-solution-id="33df6f3aa74afbe4d6f541825f6e758d"
      id={1}
      isClickable={true}
      isMatched={false}
      isOption={true}
      itemText="der KÃ¤se"
      key="1OPTION"
      onOptionClick={[MockFunction]}
      order={1}
      position="OPTION"
      showTranslation={true}
      tooltipPosition="BOTTOM"
      trainerActive={true}
      translation="translation cheese"
    />
    <WithAnimations(MatchingItem)
      appearance="DEFAULT"
      data-solution-id="b9cf73d37e1dd788faa5bf3cfaf3fd44"
      id="b9cf73d37e1dd788faa5bf3cfaf3fd44"
      isClickable={true}
      isHidden={false}
      isMatched={false}
      isOption={false}
      itemText="the rhino"
      key="b9cf73d37e1dd788faa5bf3cfaf3fd44BASE"
      onOptionClick={[MockFunction]}
      order={1}
      position="BASE"
      showTranslation={true}
      tooltipPosition="BOTTOM"
      trainerActive={true}
    />
    <WithAnimations(MatchingItem)
      appearance="DEFAULT"
      data-solution-id="b9cf73d37e1dd788faa5bf3cfaf3fd44"
      id={2}
      isClickable={true}
      isMatched={false}
      isOption={true}
      itemText="das Rhinozeros"
      key="2OPTION"
      onOptionClick={[MockFunction]}
      order={1}
      position="OPTION"
      showTranslation={true}
      tooltipPosition="TOP"
      trainerActive={true}
      translation="translation rhino"
    />
  </PoseGroup>
  <styled.div />
</styled.div>
`;
